packages/cli-plugin-metro/src/tools/KeyPressHandler.ts

```diff
@@ -0,0 +1,73 @@
+import {CLIError, logger} from '@react-native-community/cli-tools';
+
+const CTRL_C = '\u0003';
+
+/** An abstract key stroke interceptor. */
+export class KeyPressHandler {
+  private isInterceptingKeyStrokes = false;
+  private isHandlingKeyPress = false;
+
+  constructor(public onPress: (key: string) => Promise<any>) {}
+
+  /** Start observing interaction pause listeners. */
+  createInteractionListener() {
+    // Support observing prompts.
+    let wasIntercepting = false;
+
+    const listener = ({pause}: {pause: boolean}) => {
+      if (pause) {
+        // Track if we were already intercepting key strokes before pausing, so we can
+        // resume after pausing.
+        wasIntercepting = this.isInterceptingKeyStrokes;
+        this.stopInterceptingKeyStrokes();
+      } else if (wasIntercepting) {
+        // Only start if we were previously intercepting.
+        this.startInterceptingKeyStrokes();
+      }
+    };
+
+    return listener;
+  }
+
+  private handleKeypress = async (key: string) => {
+    // Prevent sending another event until the previous event has finished.
+    if (this.isHandlingKeyPress && key !== CTRL_C) {
+      return;
+    }
+    this.isHandlingKeyPress = true;
+    try {
+      logger.debug(`Key pressed: ${key}`);
+      await this.onPress(key);
```

Comment:
when this gets sync (which it is right now in fact), `this.isHandlingKeyPress` gets unnecessary